From 8ed212bee495ccd9e885eea54f27d21cfc0a0c6d Mon Sep 17 00:00:00 2001
From: Mark Thomas <markt@apache.org>
Date: Thu, 30 Jan 2014 21:39:51 +0000
Subject: [PATCH] Fix DBCP-369 Fix threading issue when using multiple
 instances of the SharedPoolDataSource concurrently. I can't reproduce this
 but the issue can be seeing from looking at the code.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/dbcp/trunk@1562988 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                                |  4 ++++
 .../dbcp2/datasources/InstanceKeyObjectFactory.java    | 10 ++++------
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index fde2265860..e772754a2f 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -95,6 +95,10 @@ The <action> type attribute can be add,update,fix,remove.
         Allow accessToUnderlyingConnectionAllowed to be configured when
         configuration takes place via JNDI in a JavaEE container.
       </action>
+      <action dev="markt" issue="DBCP-369" type="fix" due-to="Michael Pradel">
+        Fix threading issue when using multiple instances of the
+        SharedPoolDataSource concurrently.
+      </action>
     </release>
     <release version="1.5.1" date="TBD" description="TBD">
       <action dev="markt" issue="DBCP-400" type="fix">
diff --git a/src/main/java/org/apache/commons/dbcp2/datasources/InstanceKeyObjectFactory.java b/src/main/java/org/apache/commons/dbcp2/datasources/InstanceKeyObjectFactory.java
index 63c404a699..2cce68dabb 100644
--- a/src/main/java/org/apache/commons/dbcp2/datasources/InstanceKeyObjectFactory.java
+++ b/src/main/java/org/apache/commons/dbcp2/datasources/InstanceKeyObjectFactory.java
@@ -14,7 +14,6 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package org.apache.commons.dbcp2.datasources;
 
 import java.io.ByteArrayInputStream;
@@ -23,10 +22,10 @@
 
 import java.util.Hashtable;
 import java.util.Map;
-import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map.Entry;
 import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
 
 import javax.naming.Context;
 import javax.naming.Name;
@@ -40,11 +39,10 @@
  *
  * @version $Revision$ $Date$
  */
-abstract class InstanceKeyObjectFactory
-    implements ObjectFactory
-{
+abstract class InstanceKeyObjectFactory implements ObjectFactory {
+
     private static final Map<String, InstanceKeyDataSource> instanceMap =
-            new HashMap<>();
+            new ConcurrentHashMap<>();
 
     synchronized static String registerNewInstance(InstanceKeyDataSource ds) {
         int max = 0;
